vlasovsolver/vlasovmover.cpp

```diff
@@ -272,47 +272,52 @@ void calculateAcceleration(const uint popID,const uint globalMaxSubcycles,const
-   for (size_t c=0; c<propagatedCells.size(); ++c) {
-      const CellID cellID = propagatedCells[c];
-      const Real maxVdt = mpiGrid[cellID]->get_max_v_dt(popID);
-      
-      //compute subcycle dt. The length is maxVdt on all steps
-      //except the last one. This is to keep the neighboring
-      //spatial cells in sync, so that two neighboring cells with
-      //different number of subcycles have similar timesteps,
-      //except that one takes an additional short step. This keeps
-      //spatial block neighbors as much in sync as possible for
-      //adjust blocks.
-      Real subcycleDt;
-      if( (step + 1) * maxVdt > dt) {
-         subcycleDt = max(dt - step * maxVdt, 0.0);
-      } else{
-         subcycleDt = maxVdt;
-      }
-
-      //generate pseudo-random order which is always the same irrespective of parallelization, restarts, etc.
-      char rngStateBuffer[256];
-      random_data rngDataBuffer;
-
-      // set seed, initialise generator and get value. The order is the same
-      // for all cells, but varies with timestep.
-      memset(&(rngDataBuffer), 0, sizeof(rngDataBuffer));
-      #ifdef _AIX
-         initstate_r(P::tstep, &(rngStateBuffer[0]), 256, NULL, &(rngDataBuffer));
-         int64_t rndInt;
-         random_r(&rndInt, &rngDataBuffer);
-      #else
-         initstate_r(P::tstep, &(rngStateBuffer[0]), 256, &(rngDataBuffer));
-         int32_t rndInt;
-         random_r(&rngDataBuffer, &rndInt);
-      #endif
+   // Semi-Lagrangian acceleration for those cells which are subcycled,
+   // dimension-by-dimension
+   for (int order_step=0; order_step<3; order_step++) {
+      #pragma omp parallel for schedule(dynamic,1)
+      for (size_t c=0; c<propagatedCells.size(); ++c) {
+         const CellID cellID = propagatedCells[c];
+         const Real maxVdt = mpiGrid[cellID]->get_max_v_dt(popID);
          
-      uint map_order=rndInt%3;
-      phiprof::start("cell-semilag-acc");
-      cpu_accelerate_cell(mpiGrid[cellID],popID,map_order,subcycleDt);
-      phiprof::stop("cell-semilag-acc");
+         //compute subcycle dt. The length is maxVdt on all steps
+         //except the last one. This is to keep the neighboring
+         //spatial cells in sync, so that two neighboring cells with
+         //different number of subcycles have similar timesteps,
+         //except that one takes an additional short step. This keeps
+         //spatial block neighbors as much in sync as possible for
+         //adjust blocks.
+         Real subcycleDt;
+         if( (step + 1) * maxVdt > dt) {
+            subcycleDt = max(dt - step * maxVdt, 0.0);
+         } else{
+            subcycleDt = maxVdt;
+         }
+
+         //generate pseudo-random order which is always the same irrespective of parallelization, restarts, etc.
+         char rngStateBuffer[256];
+         random_data rngDataBuffer;
+
+         // set seed, initialise generator and get value. The order is the same
```

Comment:
This randomization, which provides just a single integer (same for all cells but depends randomly on timestep) could be outside the order_step loop. No point in re-randomizing the same number over and over again.

Reply:
I was wondering the same, but this random number generator is so bloody fast that it probably isn't even worth the effort moving it outside the loop. :)